rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+
+A **Link Type** is a JSON schema that satisfies the following JSON meta-schema:
+
+```json
+{
+  "$id": "https://blockprotocol.org/type-system/0.2/schema/meta/link-type",
+  "description": "Specifies the structure of a Link Type",
+  "properties": {
+    "kind": {
+      "const": "linkType"
+    },
+    "$id": {
+      "type": "string",
+      "format": "uri"
+    },
+    "name": { "type": "string" },
+    "description": { "type": "string" },
+    "relatedKeywords": {
+      "type": "array",
+      "items": { "type": "string" }
+    }
+  },
+  "additionalProperties": false,
+  "required": ["kind", "$id", "name", "description"]
+}
+```
+
+> 💭 As the next section (Entity Types) will outline, there's an additional step to quantifying the types of things on either end of the relationship, as such the descriptions **should** be very general and most of the time not contain specific words about types (such as person).
+
+**Example 1** - `Owns`
+
+```json
+{
+  "kind": "linkType",
+  "$id": "https://blockprotocol.org/types/@alice/link-type/owns",
+  "name": "Owns",
+  "description": "Have (something) as one's own; possess"
+}
+```
+
+**Example 2** - `Submitted By`
+
+```json
+{
+  "kind": "linkType",
+  "$id": "https://blockprotocol.org/types/@alice/link-type/submitted-by",
+  "name": "Submitted By",
+  "description": "Suggested, proposed, or presented by"
+}
+```
+
+## Entity Types
+
+An **Entity Type** is a JSON schema that satisfies the following JSON meta-schema:
+
+```json
+{
+  "$id": "https://blockprotocol.org/type-system/0.2/schema/meta/entity-type",
+  "description": "Specifies the structure of an Entity Type",
+  "type": "object",
+  "properties": {
+    "kind": {
+      "const": "entityType"
+    },
+    "$id": {
+      "type": "string",
+      "format": "uri"
+    },
+    "name": { "type": "string" },
+    "description": { "type": "string" },
+    "properties": { "$ref": "#/$defs/propertyTypeObject" },
+    "required": {
+      "type": "array",
+      "items": { 
+        "type": "string",
+        "format": "uri"
+      }
+    },
+    "requiredLinks": {
+      "$comment": "A list of link-types which are required. This is a separate field to 'required' to avoid breaking standard JSON schema validation",
+      "type": "array",
+      "items": { "type": "string" }
+    },
+    "links": {
+      "type": "object",
+      "propertyNames": {
+        "$comment": "Property names must be a valid URI to a link-type",
+        "type": "string",
+        "format": "uri"
+      },
+      "patternProperties": {
+        ".*": {
+          "type": "object",
+          "oneOf": [
+            {
+              "properties": {
+                "ordered": { "type": "boolean", "default": false },
+                "type": { "const": "array" }
+              },
+              "required": ["ordered", "type"]
+            },
+            {}
+          ],
+          "additionalProperties": false
+        }
+      }
+    }
+  },
+  "additionalProperties": false,
```

Comment:
It's only half intentional. I think the intention was somewhat that we'd rediscover some of these in implementation and update as necessary. Inheritance and such specifically we know needs a different RFC / explanation somewhere, so rather than blocking this on that it seemed sensible to update these meta-schemas in that explanation.

We should probably add `default` and `examples` in this RFC though.

> We may also wish to include $schema with a const on each type, to point to our meta-schema for that type.

This is something we needed to look into a little further as JSON schema documentation on `$schema` is unclear, and they often have them as self-describing in tutorials (i.e. the `$schema` tag points to `$id`, _not just for the official drafts of JSON schema_), the intention and hope was to do as you suggest, and I think we can.

Reply:
Fair enough – we don't talk about inheritance in the current spec, so I think it's fine to leave out for now and handle properly when we figure it out.

Reply:
I think it makes a lot of sense to add both `default` and `examples`. I do have some questions, however:

For both, how do we handle `link`s? 
In this proposal, they are co-located next to the `properties`, but in BP we would split them up. Would we then only allow `examples` and `default` to describe the `properties` part of an Entity Type? Specifying a default for an entity type that requires a link seems impossible

Reply:
Yes, they would only provide an example value for `properties`. Dummy data involving _other_ records (whether types, other entities, or links) can be provided via `example-graph.json` - [spec](https://blockprotocol.org/docs/spec/graph-service-specification#block-package) - [example](https://github.com/hashintel/hash/blob/main/packages/blocks/table/example-graph.json)

Reply:
I've added `default` and `examples` to the spec here https://github.com/blockprotocol/blockprotocol/pull/352/commits/75d3277cc5a05689bad0f80bde0bfa9facb40f5c
For now I haven't added any example usages in the RFC or written about it. I think that's fine for now.